README.md
OCBNET-SourceMap-XS 0.01
========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2015 by Marcel Greter

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.22.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension OCBNET::SourceMap::XS.

0.01  Fri Aug 14 18:56:15 2015
	- Initial version created



lib/OCBNET/SourceMap/XS.pm
package OCBNET::SourceMap::XS;

use 5.008000;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use XS ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(

) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);

our $VERSION = '0.01';

require XSLoader;
XSLoader::load();

package OCBNET::SourceMap::XS::LineMap;

# cannot get this to work directly in XS with the OVERLOAD keyword?
use overload fallback => 1, '==' => \&OCBNET::SourceMap::XS::LineMap::equals;

1;
__END__
###############################################################################
###############################################################################

=head1 NAME

OCBNET::SourceMap::XS - Perl extension for source-map manipulation

=head1 SYNOPSIS

  use OCBNET::SourceMap::XS;

  use File::Slurp qw(read_file);
  # read json file to create a sourcemap document
  my $json_file = 'deps/srcmap/test/output.js.map';
  my $json = read_file( $json_file, binmode => ':utf8' );
  my $srcmap = new OCBNET::SourceMap::XS::SrcMapDoc($json);

=head1 CLASSES

  my $srcmap = new OCBNET::SourceMap::XS::SrcMapDoc();
  my $mappings = new OCBNET::SourceMap::XS::Mappings(version);
  my $linemap = new OCBNET::SourceMap::XS::LineMap();
  # 0: col, 1: src_idx, 2: src_line, 3: src_col, 4: tkn_idx
  my $colmap = new OCBNET::SourceMap::XS::ColMap(0,1,2,3,4);

=head1 DESCRIPTION

API to load, create and manipulate SourceMaps as used by javascript
and css utilities and processors. The whole data is processed by a
C++ implementation, which is interfaced via XS for optimal performance.
Provides certain ultilities that might be usefull if you want to create
your own source-maps or need to alter or work with existing ones.
Currently we only support the most common L<V3 format|https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit>, but so far I have
not seen anything using any other format.

=head2 EXPORT

None by default.

=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

GitHub: L<https://github.com/mgreter/OCBNET-SourceMap-XS>

=head1 AUTHOR

Marcel Greter, E<lt>perl-libsass@ocbnet.chE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by Marcel Greter

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.22.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


Makefile.PL
use 5.008000;
use Getopt::Long;
use ExtUtils::CppGuess;
use ExtUtils::MakeMaker;

# TIP: use `dmake -P#jobs` on windows

# See lib/ExtUtils/MakeMaker.pm for details of how to
# influence content of the Makefile that is written.

use strict;
use warnings;

my $debugmode = 0;
my $compiler = undef;

# arrays for various switches
my (@libs, @flags, @defs, @incs);

# start by adding the main incs
push @incs, 'deps/srcmap/src',
            'deps/srcmap/deps/json',
            'deps/srcmap/deps/cencode';

GetOptions(
	'--compiler:s' => \$compiler,
	'--build-debug' => \$debugmode,
);

warn "Use <$compiler> as compiler" if $compiler;
die "Debug is not yet implemented" if $debugmode;

my $guess = ExtUtils::CppGuess->new(
	(defined($compiler) ? (cc => $compiler) : ()),
);

if ($guess->is_gcc) { print "Detected gcc compiler...\n"; }
elsif ($guess->is_msvc) { print "Detected MSVC compiler...\n"; }
else { print "Unknown compiler, trying anyway...\n"; }

# set c++11 flags
if ($guess->is_gcc) {
	# push @flags, '-xc++'; # also from CppGuess
	# need version which is not given via CppGuess
	my $gcc_version = `$guess->{cc} --version`;
	if ($gcc_version =~m/([4-9]\d*)\.(\d+)\.(\d+)/) {
		if ($1 >= 4 && $2 >= 7) {
			push @defs, "USE_STD_SHARED_PTR";
			push @flags, '-std=c++0x';
		}
		elsif ($1 >= 4 && $2 >= 4) {
			push @defs, "USE_STD_SHARED_PTR";
			push @flags, '-std=c++0x';
		}
		elsif ($1 == 4 && $2 >= 2) {
			push @defs, "USE_STD_TR1_SHARED_PTR";
			push @flags, '-std=gnu++0x';
		}
	}
# probably MSVC
} else {
	# fallback to more common tr1 implementation
	# not sure if that works, but try it anyways
	push @defs, "USE_STD_TR1_SHARED_PTR";
}

# add libgomp for performance timings
push @libs, '-lgomp' if ($guess->is_gcc);

# not sure why this does not work otherwise
push @flags, '-o $*.o' if ($guess->is_gcc);

# add our custom compiler flags to CppGuess
$guess->add_extra_linker_flags(join(' ', @libs));
$guess->add_extra_compiler_flags(join(' ', @flags));

# get the compiler flags generated by CppGuess
my %compiler_flags = $guess->makemaker_options();

# remove c++ flag (only needed for cpp files)
# $compiler_flags{'CCFLAGS'} =~ s/\-xc\+\+//g;

# cleanup some unnecessary whitespace
$compiler_flags{'CCFLAGS'} =~ s/^\s+//g;
$compiler_flags{'CCFLAGS'} =~ s/\s+$//g;
$compiler_flags{'CCFLAGS'} =~ s/\s+/ /g;

WriteMakefile(
	LICENSE            => 'perl',
	NAME               => 'OCBNET::SourceMap::XS',
	VERSION_FROM       => 'lib/OCBNET/SourceMap/XS.pm',
	ABSTRACT_FROM      => 'lib/OCBNET/SourceMap/XS.pm',
	AUTHOR             => q{Marcel Greter <perl-libsass@ocbnet.ch>},
	# runtime dependencies
	PREREQ_PM          => {
		'perl'                  => 5.008000,
		'Carp'                  => 1.01,
		'version'               => 0,
		'warnings'              => 0,
		'strict'                => 0,
	},
	# build dependencies
	BUILD_REQUIRES     => {
		'Getopt::Long'         => 0.01,
		'ExtUtils::CppGuess'   => 0.09,
		'ExtUtils::MakeMaker'  => 6.52,
	},
	# test dependencies
	# TEST_REQUIRES      => { },
	# config dependencies (???)
	CONFIGURE_REQUIRES => {
		'Getopt::Long'         => 0.01,
		'ExtUtils::CppGuess'   => 0.09,
		'ExtUtils::MakeMaker'  => 6.52,
	},
	# options are set by CppGuess
	# LIBS               => [''],
	# CCFLAGS            => '',
	$guess->makemaker_options(),
	# LDDLFLAGS          => '',
	INC                => join(" ", map { sprintf '-I%s', $_ } @incs),
	DEFINE             => join(" ", map { sprintf '-D%s', $_ } @defs),
	OBJECT             => [
	                        'deps/srcmap/src/document.o',
	                        'deps/srcmap/src/mappings.o',
	                        'deps/srcmap/src/map_line.o',
	                        'deps/srcmap/src/map_col.o',
	                        'deps/srcmap/src/pos_idx.o',
	                        'deps/srcmap/src/pos_txt.o',
	                        'deps/srcmap/src/format/v3.o',
	                        'deps/srcmap/deps/json/json.o',
	                        'deps/srcmap/deps/cencode/cdecode.o',
	                        'deps/srcmap/deps/cencode/cencode.o',
	                        '$(O_FILES)'
	                      ],
	TYPEMAPS           => [
	                        'xsh/sourcemap.map'
	                      ],
);
